streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java

```diff
@@ -115,6 +117,7 @@ public void run() {
 
         private void runOnce() throws InterruptedException {
             performActionsOnTasks();
+            initializeTasksIfNeeded();
```

Comment:
I wanted to avoid keeping the `tasksAndActionsLock` for too long. Based on what you said initialization can take some time, so there is some potential for blocking the main thread here. This is more or less just defensive programming and I don't have a very concrete situation where this becomes a problem. One can imagine that if you get two rebalances in a row, the second rebalance will be blocked in `handleAssignment` if the initialization from the first rebalance is still blocking `tasksAndActionsLock`.